src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java

```diff
@@ -0,0 +1,909 @@
+    try {
+      baseRequest.setParser(new JsonObjectParser(jsonFactory));
+      baseRequest.setResponseInterceptor(interceptor);
+      response = baseRequest.execute();
+      jsonFactory.createJsonParser(response.getContent(), Charsets.UTF_8)
+          .parseAndClose(parsedResponseInstance);
+    } catch (Exception e) {
+      handleError(requestIdentifier, requestIdentifierDescription, e);
+    } finally {
+      disconnectQuietly(response);
+    }
+  }
+
+  private static void disconnectQuietly(HttpResponse response) {
+    if (response != null) {
+      try {
+        response.disconnect();
+      } catch (IOException ignored) {
+        // Ignored.
+      }
+    }
+  }
+
+  private static void handleError(
+      String requestIdentifier, String requestIdentifierDescription, Exception e)
+          throws FirebaseProjectManagementException {
+    String messageBody = "Error while invoking Firebase Project Management service.";
+    if (e instanceof HttpResponseException) {
+      int statusCode = ((HttpResponseException) e).getStatusCode();
+      if (ERROR_CODES.containsKey(statusCode)) {
+        messageBody = ERROR_CODES.get(statusCode);
+      }
+    }
+    throw createFirebaseProjectManagementException(
+        requestIdentifier, requestIdentifierDescription, messageBody, e);
+  }
+
+  private <T> T blockOnFutureAndHandleExceptions(
+      ApiFuture<T> future,
+      String requestIdentifier,
+      String requestIdentifierDescription,
+      String errorMessage) throws FirebaseProjectManagementException {
+    try {
+      return future.get();
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof FirebaseProjectManagementException) {
+        throw (FirebaseProjectManagementException) e.getCause();
+      }
+      throw createFirebaseProjectManagementException(
+          requestIdentifier, requestIdentifierDescription, String.format("%s.", errorMessage), e);
+    } catch (CancellationException e) {
+      throw createFirebaseProjectManagementException(
+          requestIdentifier,
+          requestIdentifierDescription,
+          String.format("%s; the operation was cancelled.", errorMessage),
+          e);
+    } catch (InterruptedException e) {
+      throw createFirebaseProjectManagementException(
+          requestIdentifier,
+          requestIdentifierDescription,
+          String.format("%s; the operation was interrupted.", errorMessage),
+          e);
+    }
+  }
+
+  private static FirebaseProjectManagementException createFirebaseProjectManagementException(
+      String requestIdentifier,
+      String requestIdentifierDescription,
+      String messageBody,
+      @Nullable Exception cause) {
+    return new FirebaseProjectManagementException(
+        String.format(
+            "%s \"%s\": %s", requestIdentifierDescription, requestIdentifier, messageBody),
+        cause);
+  }
+
+  /* Helper types. */
+
+  private interface CreateAppFromAppIdFunction<T> extends ApiFunction<String, T> {}
```

Comment:
Lets remove this. Implementing types can directly implement ApiFunction interface.